// @netlify/zip-it-and-ship-it/dist/runtimes/go/builder.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = void 0;
const path_1 = require("path");
const fs_1 = require("../../utils/fs");
const shell_1 = require("../../utils/shell");
const build = ({ destPath, mainFile, srcDir }) => __awaiter(void 0, void 0, void 0, function* () {
    const functionName = (0, path_1.basename)(srcDir);
    try {
        yield (0, shell_1.runCommand)('go', ['build', '-o', destPath, '-ldflags', '-s -w'], {
            cwd: srcDir,
            env: {
                CGO_ENABLED: '0',
                GOOS: 'linux',
            },
        });
    }
    catch (error) {
        const runtime = 'go';
        error.customErrorInfo = { type: 'functionsBundling', location: { functionName, runtime } };
        console.error(`Could not compile Go function ${functionName}:\n`);
        throw error;
    }
    const stat = yield (0, fs_1.lstat)(destPath);
    return {
        mainFile,
        name: functionName,
        srcDir,
        srcPath: destPath,
        stat,
    };
});
exports.build = build;
