// @netlify/zip-it-and-ship-it/dist/manifest.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createManifest = void 0;
const path_1 = require("path");
const process_1 = require("process");
const fs_1 = require("./utils/fs");
const MANIFEST_VERSION = 1;
const createManifest = ({ functions, path }) => __awaiter(void 0, void 0, void 0, function* () {
    const formattedFunctions = functions.map(formatFunctionForManifest);
    const payload = {
        functions: formattedFunctions,
        system: { arch: process_1.arch, platform: process_1.platform },
        timestamp: Date.now(),
        version: MANIFEST_VERSION,
    };
    yield (0, fs_1.writeFile)(path, JSON.stringify(payload));
});
exports.createManifest = createManifest;
const formatFunctionForManifest = ({ mainFile, name, path, runtime, schedule }) => ({
    mainFile,
    name,
    path: (0, path_1.resolve)(path),
    runtime,
    schedule,
});
