// @netlify/zip-it-and-ship-it/dist/runtimes/node/bundlers/nft/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const nft_1 = require("@vercel/nft");
const resolve_dependency_1 = __importDefault(require("@vercel/nft/out/resolve-dependency"));
const minimatch_1 = __importDefault(require("minimatch"));
const unixify_1 = __importDefault(require("unixify"));
const fs_1 = require("../../../../utils/fs");
const base_path_1 = require("../../utils/base_path");
const included_files_1 = require("../../utils/included_files");
const es_modules_1 = require("./es_modules");
// Paths that will be excluded from the tracing process.
const ignore = ['node_modules/aws-sdk/**'];
const appearsToBeModuleName = (name) => !name.startsWith('.');
const bundle = ({ basePath, config, featureFlags, mainFile, pluginsModulesPath, repositoryRoot = basePath, }) => __awaiter(void 0, void 0, void 0, function* () {
    const { includedFiles = [], includedFilesBasePath } = config;
    const { exclude: excludedPaths, paths: includedFilePaths } = yield (0, included_files_1.getPathsOfIncludedFiles)(includedFiles, includedFilesBasePath || basePath);
    const { paths: dependencyPaths, rewrites } = yield traceFilesAndTranspile({
        basePath: repositoryRoot,
        config,
        mainFile,
        pluginsModulesPath,
        transpile: featureFlags.nftTranspile,
    });
    const filteredIncludedPaths = (0, included_files_1.filterExcludedPaths)([...dependencyPaths, ...includedFilePaths], excludedPaths);
    const dirnames = filteredIncludedPaths.map((filePath) => (0, path_1.normalize)((0, path_1.dirname)(filePath))).sort();
    // Sorting the array to make the checksum deterministic.
    const srcFiles = [...filteredIncludedPaths].sort();
    return {
        basePath: (0, base_path_1.getBasePath)(dirnames),
        inputs: dependencyPaths,
        mainFile,
        rewrites,
        srcFiles,
    };
});
const ignoreFunction = (path) => {
    const normalizedPath = (0, unixify_1.default)(path);
    const shouldIgnore = ignore.some((expression) => (0, minimatch_1.default)(normalizedPath, expression));
    return shouldIgnore;
};
const traceFilesAndTranspile = function ({ basePath, config, mainFile, pluginsModulesPath, transpile, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const fsCache = {};
        const { fileList: dependencyPaths, esmFileList, reasons, } = yield (0, nft_1.nodeFileTrace)([mainFile], {
            base: basePath,
            ignore: ignoreFunction,
            readFile: (path) => __awaiter(this, void 0, void 0, function* () {
                try {
                    const source = (yield (0, fs_1.cachedReadFile)(fsCache, path, 'utf8'));
                    return source;
                }
                catch (error) {
                    if (error.code === 'ENOENT' || error.code === 'EISDIR') {
                        return null;
                    }
                    throw error;
                }
            }),
            resolve: (specifier, parent, ...args) => __awaiter(this, void 0, void 0, function* () {
                try {
                    return yield (0, resolve_dependency_1.default)(specifier, parent, ...args);
                }
                catch (error) {
                    // If we get a `MODULE_NOT_FOUND` error for what appears to be a module
                    // name, we try to resolve it a second time using `pluginsModulesPath`
                    // as the base directory.
                    if (error.code === 'MODULE_NOT_FOUND' && pluginsModulesPath && appearsToBeModuleName(specifier)) {
                        const newParent = (0, path_1.join)(pluginsModulesPath, (0, path_1.basename)(parent));
                        return yield (0, resolve_dependency_1.default)(specifier, newParent, ...args);
                    }
                    throw error;
                }
            }),
        });
        const normalizedDependencyPaths = [...dependencyPaths].map((path) => basePath ? (0, path_1.resolve)(basePath, path) : (0, path_1.resolve)(path));
        const rewrites = transpile
            ? yield (0, es_modules_1.transpileESM)({ basePath, config, esmPaths: esmFileList, fsCache, reasons })
            : undefined;
        return {
            paths: normalizedDependencyPaths,
            rewrites,
        };
    });
};
const getSrcFiles = function ({ basePath, config, mainFile }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { includedFiles = [], includedFilesBasePath } = config;
        const { exclude: excludedPaths, paths: includedFilePaths } = yield (0, included_files_1.getPathsOfIncludedFiles)(includedFiles, includedFilesBasePath);
        const { fileList: dependencyPaths } = yield (0, nft_1.nodeFileTrace)([mainFile], { base: basePath, ignore: ignoreFunction });
        const normalizedDependencyPaths = [...dependencyPaths].map((path) => basePath ? (0, path_1.resolve)(basePath, path) : (0, path_1.resolve)(path));
        const includedPaths = (0, included_files_1.filterExcludedPaths)([...normalizedDependencyPaths, ...includedFilePaths], excludedPaths);
        return includedPaths;
    });
};
const bundler = { bundle, getSrcFiles };
exports.default = bundler;
